lib/pbench/server/cache_manager.py

```diff
@@ -430,29 +429,49 @@ def get_info(self, path: Path) -> JSONOBJECT:
         return fd_info
 
     @staticmethod
-    def extract(tarball_path: Path, path: str) -> Optional[str]:
-        """Extract a file from the tarball and return it as a string
+    def extract(tarball_path: Path, path: Path) -> Optional[IO[bytes]]:
+        """Returns the file stream which yields the contents of
+        a file at the specified path in the Tarball
+
+        Args:
+            tarball_path: absolute path of the tarball
+            path: relative path within the tarball
+
+        Raise:
+            BadDirpath on failure extracting the file from tarball
+        """
+        try:
+            return tarfile.open(tarball_path, "r:*").extractfile(path)
+        except Exception as exc:
+            raise BadDirpath(
+                f"A problem occurred processing {str(path)!r} from {str(tarball_path)!r}: {exc}"
+            )
 
-        Report failures by raising exceptions.
+    def filestream(self, path: str) -> Optional[dict]:
+        """Returns a dictionary containing information about the target
+        file and file stream
 
         Args:
             path: relative path within the tarball of a file
 
         Raises:
-            MetadataError on failure opening the tarball
             TarballUnpackError on failure to extract the named path
 
         Returns:
-            The named file as a string
+            Dictionary with file info and file stream
         """
-        try:
-            tar = tarfile.open(tarball_path, "r:*")
-        except Exception as exc:
-            raise MetadataError(tarball_path, exc) from exc
-        try:
-            return tar.extractfile(path).read().decode()
-        except Exception as exc:
-            raise TarballUnpackError(tarball_path, f"Unable to extract {path}") from exc
+        file_path = Path(self.name) / path
+        info = self.get_info(file_path)
+        info["stream"] = None
+
+        if info["type"] == CacheType.FILE or file_path.name == self.name:
+            try:
+                info["stream"] = Tarball.extract(self.tarball_path, file_path)
```

Comment:
When the request is `GET /api/v1/datasets/inventory/{dataset}/`, then `path` will (I believe...) be `None`.  With the code as it stands, this is going to result in a `TypeError` at line 463.

Moreover, if `path` is `None`, then we presumably want to do something completely different:
- we don't want to build `file_path`, because we already know where the tarball is (right?);
- we don't need or want to call `get_info()`, because we're not referring to the contents of the cached tarball; and,
- we don't need or want to set `info["stream"]` to `None` because we'll set it to the stream associated with the tarball.

So, I think we want a nice `if not path` before line 463, with a special case for the tarball itself.

Reply:
file_path is not building the tarball path.. it is just appending the tarball name before the passed target, like: `target="f1.json"`, it will append `file_name="dir_name/f1.json"`

Reply:
if we don't call `get_info()` then how do we get to know whether it is a file or directory?

Reply:
If `path` is false-y, then we know that the target is the tarball itself, and we know that it is a file, and we can fake whatever of the other file-info key/values that we need.  (Calling `get_info()` wouldn't help, since, as Dave pointed out, the tarball is not inside the tarball.  🙂)

Reply:
Continued below.